python/grass/pygrass/vector/table.py

```diff
@@ -714,9 +714,11 @@ def _get_driver(self):
         return decode(self.c_fieldinfo.contents.driver)
 
     def _set_driver(self, driver):
-        if driver not in ("sqlite", "pg"):
-            str_err = "Driver not supported, use: %s." % ", ".join(DRIVERS)
-            raise TypeError(str_err)
+        if driver not in DRIVERS:
+            raise NotImplementedError(
```

Comment:
From the Python doc, it actually seems to me that ValueError is more appropriate. NotImplementedError is related to abstract and unfinished class and it does not mention missing parts of user code. Also, will I get this if I pass `non-sense-drive-name`? ValueError seems quite appropriate then. Unless there is a good reason for NotImplementedError, I think this needs to be ValueError.

> [exception ValueError](https://docs.python.org/3/library/exceptions.html#ValueError)
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError).

> [exception NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError)
> This exception is derived from [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError). In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

Reply:
If a driver is supported by GRASS but not by PyGRASS, then `NotImplementedError` is appropriate ("your value is valid in theory but we haven't implemented this feature"). I will raise it for `ogr` and `dbf` only. Otherwise, indeed, `ValueError` would be correct. One could make a case for raising further a `TypeError` for non-string arguments but I think this would be over-engineered and possibly unpythonic.

(As far as I know abstract base methods are supposed to just `pass`; an exception for them is raised by Python.)

Reply:
Do you have some examples or best practices which show `NotImplementedError` should be used here? The use case in Python doc for `NotImplementedError` mentions two one use cases and that is abstract methods and not yet implemented methods, although, as these two SO/SE posts mention (https://stackoverflow.com/a/44316063/574907, https://softwareengineering.stackexchange.com/a/231398/221122), it does not explicitly exclude other use cases, the focus seems to be on methods (and both posts also state that in the context of todos). I think the `pass` you are thinking about applies to defining methods using the *abc* package.

I searched for "NotImplementedError + backend" to get some examples. In Celery, `NotImplementedError` is really used for methods which are not implemented. Not necessarily todo, but not implemented methods which are there to satisfy interface.

It seems that [usb/backend/libusb1](https://github.com/pyusb/pyusb/blob/master/usb/backend/libusb1.py#L600) uses that outside of context of methods, although the actual context is hard to tell for me.

On the other hand, [pyca/cryptography](https://github.com/pyca/cryptography/commit/3225358d1581fc5041663895aa33898db20b0a5e) had `NotImplementedError` there and then they changed it to ValueError.

I like the last change with two different errors and I agree with not raising type error. Given the above, I think it should be two value errors with different messages. However, I'm willing to merge it as is. Let me know what you think about it.